File: frontend/src/types/reservation.ts
export enum ReservationStatus {
  PENDING = "pending",
  ACTIVE = "active",
  COMPLETED = "completed",
  CANCELLED = "cancelled",
  EXPIRED = "expired"
}

export enum ReservationPriority {
  LOW = "low",
  NORMAL = "normal",
  HIGH = "high",
  CRITICAL = "critical"
}

export interface Reservation {
  id: number;
  user_id: number;
  target_id: number;
  policy_id?: number;
  start_time: string;
  end_time: string;
  status: ReservationStatus;
  priority: ReservationPriority;
  is_recurring: boolean;
  recurrence_pattern?: any;
  is_admin_override: boolean;
  override_reason?: string;
  created_at: string;
  updated_at?: string;
  last_accessed_at?: string;
}

export interface ReservationWithDetails extends Reservation {
  target_name: string;
  target_type: string;
  user_username: string;
}

export interface ReservationPolicy {
  id: number;
  name: string;
  description?: string;
  max_duration_minutes: number;
  cooldown_minutes: number;
  max_reservations_per_day: number;
  max_reservation_days_in_advance: number;
  priority_level: number;
  allowed_device_types?: string[];
  allowed_roles?: string[];
  auto_expire_enabled: boolean;
  auto_expire_minutes: number;
  notification_before_start_minutes: number;
  notification_before_end_minutes: number;
  created_at: string;
  updated_at?: string;
}

export interface ReservationSuggestion {
  target_id: number;
  target_name: string;
  device_type: string;
  start_time: string;
  end_time: string;
  reason: string;
  score: number;
}

export interface AvailabilityResponse {
  available: boolean;
  reason: string;
  conflicts: ReservationConflict[];
}

export interface ReservationConflict {
  id: number;
  user: string;
  start_time: string;
  end_time: string;
  priority: ReservationPriority;
  is_admin_override: boolean;
}

File: backend/schemas/reservation.py
from pydantic import BaseModel, Field, validator
from typing import Optional, Dict, Any, List
from datetime import datetime
from ..models.reservation import ReservationStatus, ReservationPriority

# Base Reservation Schema
class ReservationBase(BaseModel):
    target_id: int
    start_time: datetime
    end_time: datetime
    priority: Optional[ReservationPriority] = ReservationPriority.NORMAL

    @validator('end_time')
    def end_time_must_be_after_start_time(cls, v, values):
        if 'start_time' in values and v <= values['start_time']:
            raise ValueError('end_time must be after start_time')
        return v

# Schema for creating a new reservation
class ReservationCreate(ReservationBase):
    policy_id: Optional[int] = None
    is_recurring: Optional[bool] = False
    recurrence_pattern: Optional[Dict[str, Any]] = None
    override_reason: Optional[str] = None

# Schema for updating a reservation
class ReservationUpdate(BaseModel):
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: Optional[ReservationStatus] = None
    priority: Optional[ReservationPriority] = None
    policy_id: Optional[int] = None
    is_recurring: Optional[bool] = None
    recurrence_pattern: Optional[Dict[str, Any]] = None
    is_admin_override: Optional[bool] = None
    override_reason: Optional[str] = None

    @validator('end_time')
    def end_time_must_be_after_start_time(cls, v, values, **kwargs):
        if v and 'start_time' in values and values['start_time'] and v <= values['start_time']:
            raise ValueError('end_time must be after start_time')
        return v

# Schema for reservation response
class ReservationResponse(ReservationBase):
    id: int
    user_id: int
    status: ReservationStatus
    policy_id: Optional[int] = None
    is_recurring: bool
    recurrence_pattern: Optional[Dict[str, Any]] = None
    is_admin_override: bool
    override_reason: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    last_accessed_at: Optional[datetime] = None

    class Config:
        orm_mode = True

# Schema for reservation with target and user details
class ReservationWithDetails(ReservationResponse):
    target_name: str
    target_type: str
    user_username: str

    class Config:
        orm_mode = True
